static/app/components/nav/utils.tsx

```diff
@@ -0,0 +1,170 @@
+  label: string;
+  /** Optionally, props which should be passed to a wrapping `<Feature>` guard */
+  feature?: Omit<ComponentProps<typeof Feature>, 'children'>;
+}
+
+/** NavItems are displayed in either `main` and `footer` sections */
+export interface NavItemLayout<Item extends NavItem> {
+  main: Item[];
+  footer?: Item[];
+}
+
+/** SidebarItem is a top-level NavItem which is always displayed in the app sidebar */
+export interface NavSidebarItem extends NavItem {
+  /** The icon to render in the sidebar */
+  icon: JSX.Element;
+  /** Optionally, the submenu items to display when this SidebarItem is active */
+  submenu?: NavSubmenuItem[] | NavItemLayout<NavSubmenuItem>;
+  /**
+   * The pathname (including `search` params) to navigate to when the item is clicked.
+   * Defaults to the `to` property of the first `SubmenuItem` if excluded.
+   */
+  to?: string;
+}
+
+/** SubmenuItem is a secondary NavItem which is only displayed when its parent SidebarItem is active */
+export interface NavSubmenuItem extends NavItem {
+  /**
+   * The pathname (including `search` params) to navigate to when the item is clicked.
+   */
+  to: string;
+}
+
+export type NavConfig = NavItemLayout<NavSidebarItem>;
+
+export type NavigationItemStatus = 'inactive' | 'active' | 'active-parent';
+
+/** Determine if a given SidebarItem or SubmenuItem is active */
+export function isNavItemActive(
+  item: NavSidebarItem | NavSubmenuItem,
+  location: ReturnType<typeof useLocation>
+): boolean {
+  const to = resolveNavItemTo(item);
+  if (!to) {
+    return false;
+  }
+
+  /**
+   * Issue submenu is special cased because it is matched based on query params
+   * rather than the pathname.
+   */
+  if (
+    location.pathname.includes('/issues/') &&
+    to.includes('/issues/') &&
+    to.includes('query=')
+  ) {
+    if (location.search) {
+      return hasMatchingQueryParam({to, label: item.label}, location);
+    }
+    return item.label === 'All';
+  }
+
+  const normalizedTo = normalizeUrl(to);
+  const normalizedCurrent = normalizeUrl(location.pathname);
+  // Shortcut for exact matches
+  if (normalizedTo === normalizedCurrent) {
+    return true;
+  }
+  // Fallback to legacy nav logic
+  return isItemActive({to, label: item.label});
+}
+
+export function isSubmenuItemActive(
+  item: NavSidebarItem,
+  location: ReturnType<typeof useLocation>
+): boolean {
+  if (!item.submenu) {
+    return false;
+  }
+  if (isNonEmptyArray(item.submenu)) {
+    return item.submenu.some(subitem => isNavItemActive(subitem, location));
+  }
+  return (
+    item.submenu.main.some(subitem => isNavItemActive(subitem, location)) ||
+    item.submenu.footer?.some(subitem => isNavItemActive(subitem, location)) ||
+    false
+  );
+}
+
+/** Creates a well-formed `LocationDescriptor` from a partial URL string that may contain search params */
+export function makeLocationDescriptorFromTo(to: string): LocationDescriptor {
```

Comment:
Just to be extra safe, I think we should `if(!URL.canParse(to)) throw new Error` so we dont end up with broken links somewhere without knowing :D

Reply:
Ah, I looked into this and unfortunately `URL.canParse()` doesn't help here—it assumes a fully formed URL with scheme, so a relative or absolute path won't pass. 

If we detect that and try to add a scheme ourselves, almost anything will be treated as valid.

Reply:
Ahh, I see! Yeah, makes sense in that case, the unknown url route still sends an issue to Sentry so we should be covered here, thanks for the context!